test/libsolidity/SemanticTest.cpp

```diff
@@ -88,9 +88,15 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 	if (m_runWithEwasm && !m_supportsEwasm)
 		m_runWithEwasm = false;
 
-	m_runWithABIEncoderV1Only = m_reader.boolSetting("ABIEncoderV1Only", false);
-	if (m_runWithABIEncoderV1Only && !solidity::test::CommonOptions::get().useABIEncoderV1)
-		m_shouldRun = false;
+	if (!solidity::test::CommonOptions::get().useABIEncoderV1)
+	{
+		for (auto const& [filename,source]: m_reader.sources().sources)
+			if (source.find("pragma abicoder v1;") != string::npos)
+			{
+				m_shouldRun = false;
+				break;
+			}
+	}
```

Comment:
I think this change silently breaks all multi-file tests that mix both v1 and v2 but don't have an explicit pragma for v2 (and even if they do, you said you're planning to remove the explicit v2 pragmas anyway). With `--abiencoderv1` all the files in such a test will be using v1 defeating the purpose of the test. Without `--abiencoderv1` the test will be skipped because it contains the v1 pragma.

It also results in tests that had the v1 pragma but not `ABIEncoderV1Only: true` being skipped when not running with `--abiencoderv1`. I think this is only necessary for the new codegen but you explicitly added `compileViaYul: false` to such tests so there's no need to skip them here at all now.

Actually `ABIEncoderV1Only` looks to me like a workaround for not having an explicit v1 pragma in the past. Now that we it, the whole mechanism is obsolete - does not need to be replaced with anything. Adding v1/v2 pragmas and `compileViaYul` in tests like you did is enough.

Reply:
Good observation! I actually think that an explicit "run v2 coder only" option is useful.

Reply:
@axic looks like this needs some discussion

Reply:
@cameel I think this is blocked on your comment here. I'm confused, the first part seems to be agains this PR but the second part seems to be pro?

Reply:
I'm for the PR but against this particular change. I think it could make some of the multi-source tests testing both v1 and v2 ineffective. And it is not necessary to achieve the effect you want.

Reply:
I'm sorry I just don't fully understand. Can you give some example cases?

Reply:
Tests from https://github.com/ethereum/solidity/tree/v0.8.4/test/libsolidity/syntaxTests/abiEncoder.
For example this one: https://github.com/ethereum/solidity/blob/v0.8.4/test/libsolidity/syntaxTests/abiEncoder/v2_v1_v2_modifier_sandwich.sol.

That's of course if you removed the v2 pragma from them (which you did not do yet) but with this change removed it would not matter.

Reply:
Ah, sorry that one is a syntax error. So this one: https://github.com/ethereum/solidity/blob/v0.8.4/test/libsolidity/syntaxTests/abiEncoder/v1_v2_v1_modifier_mix.sol.

Reply:
That is a syntax test. This PR should only affect semantic tests.

Reply:
Well, sure. But we could have a semantic test like this and this change would break it.

Let's put it differently: why do you think this change is necessary at all?

Reply:
Honestly I am not sure what is the hold up :wink:

The main motivation for this change is to reduce the complexity of the test suite. Before this (i.e. the time I was preparing this change) it takes quite a bit of time understanding all the relationships between abicoder, yul, etc. and what combinations are allowed, are accepted silently, etc.

Reply:
Maybe what is unclear to the reader, is that `abicoder v1` and `compileViaYul: both` (or true) means that `compileViaYul` will ignore the `abicoder v1` and implicity use v2 (we display a warning).

Because of this reason, we already have certain tests mostly suffixed with `_v1.sol` which target to be run only for v1. These tests are for those cases where there is an encoding difference between v1 and v2.

This change just removes this extra `ABIEncoderV1Only: true` setting and makes it rely purely on the pragma, making the situation cleaner in my opinion. Note, currently it is even allowed to have `pragma abicoder v2;` **and** `ABIEncoderV1Only: true`.

We just have *way too many* settings in the test suite.

Reply:
This is a good description of the motivation for the PR as a whole and I agree with this. But it does not explain why you need to skip the v1 tests when `--abiencoderv1` is not present. They would still run just fine.